lib/github_changelog_generator/generator/generator.rb

```diff
@@ -92,13 +92,14 @@ def generate_entry_between_tags(older_tag, newer_tag)
       Entry.new(options).generate_entry_for_tag(filtered_pull_requests, filtered_issues, newer_tag_name, newer_tag_link, newer_tag_time, older_tag_name)
     end
 
-    # Filters issues and pull requests based on, respectively, `closed_at` and `merged_at`
-    #  timestamp fields.
+    # Filters issues and pull requests based on, respectively, `actual_date`
+    # and `merged_at` timestamp fields. `actual_date` is the detected form of
+    # `closed_at` based on merge event SHA commit times.
     #
     # @return [Array] filtered issues and pull requests
     def filter_issues_for_tags(newer_tag, older_tag)
       filtered_pull_requests = filter_by_tag(@pull_requests, newer_tag)
-      filtered_issues        = delete_by_time(@issues, "closed_at", older_tag, newer_tag)
+      filtered_issues        = delete_by_time(@issues, "actual_date", older_tag, newer_tag)
```

Comment:
be careful with that. it should be tested well. I remember, that have a reason to change `actual_date` to `closed_at` before. Need to check logs to understand why. but there is couple of tricky cases related with that string

Reply:
Okay, I can do some more looking into it. If I recall, the `actual_date` was problematic for PRs but not for issues, but that should be irrelevant due to #619 using git history for PRs now.

---

Dump of investigation thoughts follow.

Before #504, PRs and issues were filtered by `actual_date` / `actual_date`. Those values are not from the github API but actually set by the fetcher in GCG based on GitHub API events: https://github.com/github-changelog-generator/github-changelog-generator/blob/1f68ef8bad451bb7df516700acedf1f158c4564f/lib/github_changelog_generator/generator/generator_fetcher.rb#L53 (That method looks for events of "merged" for PRs or "closed" for issues and uses that for the "actual_date" of the item.)

The `merged_at` and `closed_at` values on PRs and issues do come from the GitHub API at the issue level rather than the event level.

For pull requests, the following is now irrelevant as #619 now finds PR/tag relationships based on git history, but it is helpful to have written down since I'm digging into it anyway.

For pull requests, the `merged_at`/`closed_at` (identical) time is often one second later than the date of the merge commit due to GitHub marking the PR closed one second after creating the merge commit. But GCG detects the tag time based on the tagged commit's commit time, which would place any PR whose merge commit was tagged one second later than the tag, thus not considered part of the tagged release even though it was. For this, it would be better to consider both the PR close time and the tag release time to be the merge event commit's time.

For pull requests, the `merged_at`/`closed_at` time could be more than one second different from the `actual_date` if a user uses `hub merge <url>` locally and pushes the merge commit later, but this is very rare.

For pull requests, if the events do not have a "merged" event to detect the merge commit then the `merged_at` time is used anyway.

For issues, the events are inspected for any event of "closed" and if so checked for a `commit_id` to use for `actual_date`. If the "closed" event has no `commit_id` then the `closed_at` time is used for `actual_date` and there's no difference.

For issues, if the "closed" event has a `commit_id` then the author commit time is used as the `actual_date`. This occurs when a commit message has the [magic words](https://blog.github.com/2013-01-22-closing-issues-via-commit-messages/) in it and is somewhat rare. The milestone matching is probably much more useful. This would at least be improved if the issue was associated with a release based on its commit sha as PRs are now rather than time-based. I believe this also is better to use the commit time, for example [this](https://api.github.com/repos/github-changelog-generator/github-changelog-generator/issues/228) issue being closed a day after the commit time of [the closing](https://api.github.com/repos/github-changelog-generator/github-changelog-generator/commits/a72ad326c40f47ac0c8f96b837e9a8d9feae4850) commit.

Issues just don't have a good mechanism for determining which release they go in, especially for projects with parallel branches. They should probably be determined in some order of priority like this:
1. Milestones matching the tag.
2. "Closing commit: <sha>" similar to the "Rebased commit: <sha>" comment in #619 (this doesn't exist yet)
3. closing event commit_id author time (or committer time? Hmm...)
4. The `closed_at` api field.

Numbers 3 and 4 are what `actual_date` does do, whereas `closed_at` only does the last one.

Reply:
This dump of investigation notes - a lot of valuable work. Thank you.

(The last bullet list made me think of "data organizing" vs "rendering" again. If it were possible to have a preview of how the issues/PRs/tags/releases/milestones would be presented, there _could_ be room for an interactive step, where the user could decide which Section a line should be in. If those choices were persisted and re-used, it could be neat.)